<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Calculate Nuclear Fraction Score • dropletQC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Calculate Nuclear Fraction Score">
<meta property="og:description" content="This simple R package can be used to calculate a useful QC metric 
    for single cell RNA-seq datasets generated with the 10x Genomics Chromium
    Single Cell Gene Expression platform: the nuclear fraction score. This score
    captures, for each input cell barcode, the fraction of RNA that likely 
    originated from unspliced (nuclear) pre-mRNA. Sheared cell membranes from 
    damaged cells in the input cell suspension primarily release cytoplasmic RNA
    into solution, while the nuclear envelope remains intact. As a result, RNA
    released from stressed or damaged cells consists mostly of mature 
    cytoplasmic mRNA and is relatively depleted of unspliced nuclear precursor 
    mRNA. This fact can be useful in identifying both; droplets containing only 
    ambient RNA, characterised by a low nuclear fraction score, and damaged 
    cells, which can be identified by a high score. The QC metric is calculated 
    by parsing the input barcoded BAM file for CB and RE flags added by 10x 
    Genomics' Cell Ranger software.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">dropletQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/dropletQC.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/WalterMuskovic/dropletQC/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="dropletqc" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#dropletqc" class="anchor"></a>dropletQC</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>This is a simple R package to calculate, for every requested cell barcode in a provided scRNA-seq BAM file, the nuclear fraction score:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>nuclear fraction <span class="ot">=</span> intronic reads <span class="sc">/</span> (intronic  reads  <span class="sc">+</span>  exonic  reads)</span></code></pre></div>
<p>The score captures the proportion of reads from intronic regions. These RNA fragments originate from unspliced (nuclear) pre-mRNA, hence the name “nuclear fraction”. This score can be used to help identify:</p>
<ol>
<li><p>“Empty” droplets containing ambient RNA: low nuclear fraction score and low UMI count</p></li>
<li><p>Droplets containing damaged cells: high nuclear fraction score and low UMI count</p></li>
</ol>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install dropletQC with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"WalterMuskovic/dropletQC"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="calculating-the-nuclear-fraction" class="section level2">
<h2 class="hasAnchor">
<a href="#calculating-the-nuclear-fraction" class="anchor"></a>Calculating the nuclear fraction</h2>
<p>There are two functions which can be used to calculate the nuclear fraction; <code>nuclear_fraction_tags</code> and <code>nuclear_fraction_annotation</code>.</p>
<p>If your BAM file contains region tags which identify aligned reads as intronic or exonic, such as those produced by 10x Genomics’ Cell Ranger software, then the simplest and fastest way to calculate the nuclear fraction is to point <code>nuclear_fraction_tags</code> to the directory:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://waltermuskovic.github.io/dropletQC">dropletQC</a></span><span class="op">)</span>
<span class="va">nf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/nuclear_fraction_tags.html">nuclear_fraction_tags</a></span><span class="op">(</span>
    outs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"outs"</span>, package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span>,
     tiles <span class="op">=</span> <span class="fl">1</span>, cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">nf1</span><span class="op">)</span>
<span class="co">#&gt;                    nuclear_fraction</span>
<span class="co">#&gt; AAAAGTCACTTACTTG-1        0.9032698</span>
<span class="co">#&gt; AAAAGTGGATCTCTAA-1        0.4032761</span>
<span class="co">#&gt; AAAGCAGTTACGAAGA-1        0.3957704</span>
<span class="co">#&gt; AACGACTTCAATATGT-1        0.4004525</span>
<span class="co">#&gt; AACGGCGTCATCTGGA-1        0.8845109</span>
<span class="co">#&gt; AAGCAGGGGTCGCGAA-1        0.3929376</span></code></pre></div>
<p>Alternatively, you can point <code>nuclear_fraction_annotation</code> to a gene annotation, BAM and barcode files:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nf2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/nuclear_fraction_annotation.html">nuclear_fraction_annotation</a></span><span class="op">(</span>
 annotation_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/outs/chr1.gff3"</span>,package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span>,
 bam <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/outs/possorted_genome_bam.bam"</span>,package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span>,
 barcodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/outs/filtered_feature_bc_matrix/barcodes.tsv.gz"</span>,package <span class="op">=</span> <span class="st">"dropletQC"</span><span class="op">)</span>,
 tiles <span class="op">=</span> <span class="fl">1</span>, cores <span class="op">=</span> <span class="fl">1</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">nf2</span><span class="op">)</span>
<span class="co">#&gt;                    nuclear_fraction</span>
<span class="co">#&gt; AAAAGTCACTTACTTG-1        0.9032698</span>
<span class="co">#&gt; AAAAGTGGATCTCTAA-1        0.4032761</span>
<span class="co">#&gt; AAAGCAGTTACGAAGA-1        0.3957704</span>
<span class="co">#&gt; AACGACTTCAATATGT-1        0.4004525</span>
<span class="co">#&gt; AACGGCGTCATCTGGA-1        0.8845109</span>
<span class="co">#&gt; AAGCAGGGGTCGCGAA-1        0.3929376</span></code></pre></div>
<p>This methods is more flexible, as it makes no assumptions about how your BAM file was produced - but it will take longer. Take care that the provided barcodes match the barcode structure in the BAM file.</p>
</div>
<div id="identifying-empty-droplets-and-damaged-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#identifying-empty-droplets-and-damaged-cells" class="anchor"></a>Identifying empty droplets and damaged cells</h2>
<p>Once the nuclear fraction score has been calculated, the <code>identify_empty_drops</code> and <code>identify_damaged_cells</code> functions can be used to assist in identifying each these populations. Empty or damaged cells are flagged, not removed.</p>
</div>
<div id="more-information" class="section level2">
<h2 class="hasAnchor">
<a href="#more-information" class="anchor"></a>More information</h2>
<p>For a detailed discussion see our paper published in <strong>Journal Name</strong>:</p>
<p><a href="https://www.google.com">paper title here</a></p>
<p>For more information about the functions included in the package, including tips on how to assess the nuclear fraction score using real-world examples, see the <a href="https://waltermuskovic.github.io/dropletQC/articles/dropletQC.html">package vignette</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/WalterMuskovic/dropletQC/">https://​github.com/​WalterMuskovic/​dropletQC/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/WalterMuskovic/dropletQC/issues">https://​github.com/​WalterMuskovic/​dropletQC/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Walter Muskovic <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Walter Muskovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
