<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate the nuclear fraction statistic by quantifying intron/exon overlap — nuclear_fraction_annotation • dropletQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calculate the nuclear fraction statistic by quantifying intron/exon overlap — nuclear_fraction_annotation" />
<meta property="og:description" content="This function calculates the nuclear fraction score by parsing
all of the reads in the provided BAM file and determining their overlap with
intron and exon intervals obtained from the provided annotation file. It is
a more flexible function than nuclear_fraction_tags() because it doesn't
rely on the BAM file already containing tags describing whether reads are
aligned to intronic or exonic regions, but it is also slower, because this
information needs to be calculated." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dropletQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dropletQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/WalterMuskovic/dropletQC/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate the nuclear fraction statistic by quantifying intron/exon overlap</h1>
    <small class="dont-index">Source: <a href='https://github.com/WalterMuskovic/dropletQC/blob/master/R/nuclear_fraction_annotation.R'><code>R/nuclear_fraction_annotation.R</code></a></small>
    <div class="hidden name"><code>nuclear_fraction_annotation.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates the nuclear fraction score by parsing
all of the reads in the provided BAM file and determining their overlap with
intron and exon intervals obtained from the provided annotation file. It is
a more flexible function than <code><a href='nuclear_fraction_tags.html'>nuclear_fraction_tags()</a></code> because it doesn't
rely on the BAM file already containing tags describing whether reads are
aligned to intronic or exonic regions, but it is also slower, because this
information needs to be calculated.</p>
    </div>

    <pre class="usage"><span class='fu'>nuclear_fraction_annotation</span><span class='op'>(</span>
  <span class='va'>annotation_path</span>,
  annotation_format <span class='op'>=</span> <span class='st'>"auto"</span>,
  <span class='va'>bam</span>,
  bam_index <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='va'>bam</span>, <span class='st'>".bai"</span><span class='op'>)</span>,
  <span class='va'>barcodes</span>,
  cell_barcode_tag <span class='op'>=</span> <span class='st'>"CB"</span>,
  cores <span class='op'>=</span> <span class='fu'>future</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/future/man/re-exports.html'>availableCores</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>-</span> <span class='fl'>1</span>,
  tiles <span class='op'>=</span> <span class='fl'>1000</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>annotation_path</th>
      <td><p>character, should be a character vector pointing to the
annotation file</p></td>
    </tr>
    <tr>
      <th>annotation_format</th>
      <td><p>character. Can be one of "auto", "gff3" or "gtf".
This is passed to the 'format' argument of
GenomicFeatures::makeTxDbFromGFF(). THis should generally just be left as
"auto"</p></td>
    </tr>
    <tr>
      <th>bam</th>
      <td><p>character, should be a character vector pointing to the BAM file</p></td>
    </tr>
    <tr>
      <th>bam_index</th>
      <td><p>character, the path to the input bam file index</p></td>
    </tr>
    <tr>
      <th>barcodes</th>
      <td><p>character, either a vector of barcode names or the path to a
file barcodes.tsv.gz containging the cell barcodes. If providing the cell
barcodes as a vector, make sure that the format matches the one in the BAM
file - e.g. be mindful if there are integers appended to the end of the
barcode sequence.</p></td>
    </tr>
    <tr>
      <th>cell_barcode_tag</th>
      <td><p>character, defines the BAM tage which contains the
cell barcode e.g. "CB"</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>numeric, parsing of the BAM file can be run in parallel using
furrr:future_map() with the requested number of cores. Setting <code>cores=1</code>
will cause future_map to run sequentially.</p></td>
    </tr>
    <tr>
      <th>tiles</th>
      <td><p>integer, to speed up the processing of the BAM file we can split
transcripts up into tiles and process reads in chunks, default=1000.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical, whether or not to print progress</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data.frame. Returns a data frame containing the nuclear fraction
score. THis is just the fraction of reads that are intronic:</p>
<p>nuclear fraction = # intronic reads / (# intronic reads + # of exonic reads)</p>
<p>The row names of the returned data frame will match the order and name of
the supplied barcodes.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>nf3</span> <span class='op'>&lt;-</span> <span class='fu'>nuclear_fraction_annotation</span><span class='op'>(</span>
 annotation_path <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/outs/chr1.gff3"</span>,
  package <span class='op'>=</span> <span class='st'>"dropletQC"</span><span class='op'>)</span>,
 bam <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata/outs/possorted_genome_bam.bam"</span>,
  package <span class='op'>=</span> <span class='st'>"dropletQC"</span><span class='op'>)</span>,
 barcodes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span>
 <span class='st'>"extdata/outs/filtered_feature_bc_matrix/barcodes.tsv.gz"</span>,
  package <span class='op'>=</span> <span class='st'>"dropletQC"</span><span class='op'>)</span>,
 tiles <span class='op'>=</span> <span class='fl'>1</span>, cores <span class='op'>=</span> <span class='fl'>1</span>, verbose <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "Extracting exon and intron ranges from provided annotation file:"
#&gt; [1] "/private/var/folders/cs/qxrqw1nj61gdhl_hl29y0nv00000gp/T/RtmprS5zuk/temp_libpath14ab81e52a36b/dropletQC/extdata/outs/chr1.gff3"</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>nf3</span><span class='op'>)</span>
</div><div class='output co'>#&gt;                    nuclear_fraction
#&gt; AAAAGTCACTTACTTG-1        0.9032698
#&gt; AAAAGTGGATCTCTAA-1        0.4032761
#&gt; AAAGCAGTTACGAAGA-1        0.3957704
#&gt; AACGACTTCAATATGT-1        0.4004525
#&gt; AACGGCGTCATCTGGA-1        0.8845109
#&gt; AAGCAGGGGTCGCGAA-1        0.3929376</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Walter Muskovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


