<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Parse cell barcode and region type tags in input region of provided BAM file — parse_bam_tags • DropletQC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Parse cell barcode and region type tags in input region of provided BAM file — parse_bam_tags" />
<meta property="og:description" content="This function accepts three inputs; a GRanges defining a genomic
interval, a reference to a BAM file and a vector of cell barcodes. It then
uses uses Rsamtools::scanBam() to import the provided alignment tags which
contain the cell barcode and region type (exonic or intronic). The tags are
then parsed and returned as a data frame with two integer columns; exon and
intron." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DropletQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/DropletQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Parse cell barcode and region type tags in input region of provided BAM file</h1>
    
    <div class="hidden name"><code>parse_bam_tags.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function accepts three inputs; a GRanges defining a genomic
interval, a reference to a BAM file and a vector of cell barcodes. It then
uses uses <code><a href='https://rdrr.io/pkg/Rsamtools/man/scanBam.html'>Rsamtools::scanBam()</a></code> to import the provided alignment tags which
contain the cell barcode and region type (exonic or intronic). The tags are
then parsed and returned as a data frame with two integer columns; exon and
intron.</p>
    </div>

    <pre class="usage"><span class='fu'>parse_bam_tags</span><span class='op'>(</span><span class='va'>interval</span>, <span class='va'>bam</span>, <span class='va'>bc</span>, <span class='va'>CB_tag</span>, <span class='va'>RE_tag</span>, <span class='va'>EXON_tag</span>, <span class='va'>INTRON_tag</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>interval</th>
      <td><p>a single GRanges defining a genomic interval to import reads
from</p></td>
    </tr>
    <tr>
      <th>bam</th>
      <td><p>a reference to a BAM file to pull reads from (Rsamtools::BamFile())</p></td>
    </tr>
    <tr>
      <th>bc</th>
      <td><p>a character vector of cell barcodes of interest</p></td>
    </tr>
    <tr>
      <th>CB_tag</th>
      <td><p>character, the BAM tag containing the cell barcode sequence</p></td>
    </tr>
    <tr>
      <th>RE_tag</th>
      <td><p>character, the BAM tag containing the region type</p></td>
    </tr>
    <tr>
      <th>EXON_tag</th>
      <td><p>character, the character string that defines a read as exonic</p></td>
    </tr>
    <tr>
      <th>INTRON_tag</th>
      <td><p>character, the character string that defines a read as
intronic</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data.frame returns a data frame with two integer columns; "exon" - the
number of reads that map to exons in the requested genomic interval for each
cell barcode and "intron" - the number of reads that map to introns, where
the rows are in the same order as the input character vector of cell
barcodes. If there are no reads in the interval requested, the function will
return NULL. This function was written as a helper function to be called
internally by <code><a href='nuclear_fraction_tags.html'>DropletQC::nuclear_fraction_tags</a></code> and isn't intended for more
general use.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Walter Muskovic.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


